### ISSUE TITLE
Build error when including a type and set of types

### ISSUE THREAD

**Issue by MatrixSenpai** [apollographql/apollo-ios#729]
Running into an issue with conflicting types.

I have a schema definition that follows this format:
```
type Match {
  id: ID
  name: String
  start: Date
  winner: Team
  teams: [Team]
}
type Team {
  id: ID
  name: String
}
```

When trying to define it in Apollo on the iOS client, I get the build error 'Cannot convert value of type 'Match.Winner.Type' to 'GraphQLFragment.Type'. My definitions are
```
# TeamFragment.graphql
fragment Team on Team {
  id
  name
}

# MatchFragment.graphql
fragment Match on Match {
  id
  name
  start
  winner {
    ... Team
  }
  teams {
    ... Team
  }
}
```

If I remove the definition for the `teams` field, `Team` gets created as a `GraphQL Fragment`. If I add it, `Team` is defined as a `GraphQLSelectionSet` and throws a type error.

Am I simply defining these wrong?

**Comment by MatrixSenpai** [comment]
As an addendum to this, I created two new fragments, `WinningTeam` and `ParticipatingTeams` and assigned them to `winner` and `teams`, respectively. However, those fragments are exactly the same, and this is not necessarily expected behavior. If I try to assign either of those fragments to both fields, the build fails the same way

**Comment by designatednerd** [comment]
I'll take a deeper look into this when I get back from moving, but I know we've had some issues with fragments that have the same name as the type name. 

In general I'd expect the fragment name to be something more like `TeamDetails` or `TeamIDAndName` rather than having the same exact name as the type itself.

**Comment by MatrixSenpai** [comment]
Fair enough, but either way, it produces unexpected results

**Comment by designatednerd** [comment]
Yeah agreed. I’ll see what I can do when I’m back.

**Comment by designatednerd** [comment]
OK sorry I dropped the ball on this one. @MatrixSenpai I believe that you've basically found an undocumented edge case by naming the fragment the same thing as the type. I'm updating our documentation to tell people not to do this explicitly.

**Comment by MatrixSenpai** [comment]
I've not had an issue with naming the fragments exactly the same as they are defined in the schema. The issue comes in when I try to use the same fragment in two different places as children of an object. For example, if I have the following schema
```
fragment Match on Match {
  // ...
  winner {
    ... Team
  }
  participants {
    ... Team
  }
}
```
where `winner` is of type `Team` and `participants` is of type `Array<Team>`, that's where I run into an issue

**Comment by designatednerd** [comment]
Ahh, ok. So what happens if you have:

```graphql
fragment Match on Match {
  // ...
  winner {
    ... WinningTeam
  }
  participants {
    ... Team
  }
}
```

**Comment by MatrixSenpai** [comment]
Yep, that works perfectly. It's not a _breaking_ issue, just seems a little odd that I can't use the same type 😆

**Comment by designatednerd** [comment]
Hopefully, the switch to fragments as protocols outlined in #939 should alleviate this - it's a lot easier to have two objects which confirm to the same protocol than two types with the same name at the same level.

**Comment by designatednerd** [comment]
(I am going to close out my PR without merging since I misunderstood the use case)

**Comment by MatrixSenpai** [comment]
Sounds good to me, I will test out #939 if/when it gets merged in! Until then, leaving open for further review and testing

**Comment by designatednerd** [comment]
To be clear #939 itself is an RFC, so that's not something that will be merged, but I'll try to ping you when the fragments-as-protocols stuff is actually merged.

**Comment by MatrixSenpai** [comment]
Pinging @designatednerd since it looks like #939 was closed just a few days ago

**Comment by designatednerd** [comment]
The RFC was closed and I've started work on the rewrite, but it's going to be A Process. I'll make sure to update this issue when I get to it!

**Comment by lauraweaver** [comment]
This is happening to me! Is there any possibility that this will still be worked on?

**Comment by designatednerd** [comment]
Yes, it's being worked on. I made a gigantic understatement when I said: `it's going to be A Process.`.
